### ISSUE TITLE
Can you use bearssl to replace openssl?

### ISSUE THREAD

**Issue by skyformat99** [ntop/n2n#231]
bearssl: 
https://bearssl.org/

Bearssl is much better than openssl.
Openssl has too many bugs

**Comment by Logan007** [comment]
Could you provide some more details about your concerns with openSSL?

**Comment by skyformat99** [comment]
Openssl is way too big and more cumbersome.
BearSSL is published under the MIT License,  highly portable, small.
I also want to use this library on an embedded platform so it should be small, easy, secure and free.

**Comment by lucaderi** [comment]
I think that this is a good idea actually. Anybody interested in contributing to this ticket?

**Comment by lucktu** [comment]
If bearssl is better, then why not support it?

**Comment by Logan007** [comment]
Having some time at hand these days, I volunteer to try some basic test coding to better find out how to employ bearSSL for AES encryption (documentation in terms of examples seems to be rare) and how it performs compared to openSSL.

Will report what I find.

No real objections but the following questions are intriguing me a bit: Could bearSSL be considered as reviewed and maintained in the future as openSSL?

**Comment by Logan007** [comment]
There are two interfaces through which we could communicate with bearSSL. The first one I tried is the procedural interface, it is a bit like openSSL's _non_-`evp_*` interface. The following code shows how it works (compile with `-lbearssl`):

```
#include <stdio.h>
#include <bearssl.h>


int main(int argc, char* argv[]) {
        // sizewise, the key array is prepared to take up to 256 bit keys
        uint8_t key[32] = {128,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0 };

        size_t key_len = 16;
        br_aes_big_cbcenc_keys enc_ctx;
        // generate the round keys, this context can be re-used
        br_aes_big_cbcenc_init (&enc_ctx, key, key_len);

        // data
        uint8_t data[1024] = { 128,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0 };
        size_t data_len = sizeof(data);

        // iv
        uint8_t enc_iv[16] = {128,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0 };

fprintf (stderr, "%16llx\n", *(uint64_t*)data);

        // encryption; happens in-place, i.e. data is overwritten;
        // iv not original anymore afterwards (contains last cipher block)
        br_aes_big_cbcenc_run(&enc_ctx, enc_iv, data, data_len);

fprintf (stderr, "%16llx\n", *(uint64_t*)data);

        // decryption requires its own special context / context type
        br_aes_big_cbcdec_keys dec_ctx;
        // generate the round keys, this context can be re-used
        br_aes_big_cbcdec_init (&dec_ctx, key, key_len);

        // iv
        uint8_t dec_iv[16] = {128,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0 };

        // decryption; happens in-place, i.e. data is overwritten;
        // iv not original anymore afterwards (contains last cipher block)
        br_aes_big_cbcdec_run(&dec_ctx, dec_iv, data, data_len);

// should deliver the 0EDD33... (endianess!) for (iv XOR data) = 00000... and key = 80000...
fprintf (stderr, "%16llx\n", *(uint64_t*)data);
}
```

Problem: bearSSL includes several implementations of AES for several platforms and/or purposes (`aes_big`, `aes_small`, `aes_ct`, `aes_ct64`, `aes_x86ni`, `aes_pwr8`). The name of the respective function calls and also the names of types/structs differ. So we would probably need platform-dependent compilation as well as a lot of places in the code to take care of – just replace every occurence of "...big_aes..." with maybe "...aes_x86ni..." in the code above.

A possible solution to this could be offered by bearSSL's object oriented programming-like interface (the one to try next) which would reduce the number of places to use specifically named code. __EDIT:__ _bearSSL offers some help: I found some `br_aes_x86ni_cbcenc_get_vtable()` and `br_aes_pwr8_cbcenc_get_vtable()` in the API description. Let's see how they could come in handy. But still... openSSL does take care of this by itself._

I will keep you posted about my field trials.

Speedwise, at least `aes_x86ni` looks promising (compared to openSSL's `evp_*` interface) indeed, but no definite results yet.

Any thoughts?

**Comment by Logan007** [comment]
Going through the test-code of bearSSL and some documentation of the object oriented programming-like API, I dissected and commented below the neccessary steps for how to use it. The following code performs the encryption:

```
#include <stdio.h>
#include <bearssl.h>

int main(int argc, char* argv[]) {

        // [A] specify class to use (choose cipher implementation, here: 'big')
        br_block_cbcenc_class vtable_enc = br_aes_big_cbcenc_vtable;

        // [B] instance of class (object interface, externally providing access to methods and fields)
        const br_block_cbcenc_class **enc;

        // [C1] process of instantiation
        // the actual memory (inner) representation of the object (encryption context internally also provides
        // space for round keys, hence the name), 'gen' fits all (e.g. 'big', 'small', ...)
        br_aes_gen_cbcenc_keys v_enc_ctx;
        //  [C2] connecting interface to actual instance's data structure...
        enc = &v_enc_ctx.vtable;
        // [D] ... and set default values (properties, pointers to methods) using the class' 'constructor' - otherwise object is useless
        // note: if key and key_len are already known, those could be set here. otherwise: some defaults
        vtable_enc.init(enc, 0, 0);

        // [E] from here on: use the object
        fprintf (stderr, "block length: %u\n", (**enc).block_size);

        // sizewise, the key array is prepared to take up to 256 bit keys
        uint8_t key[32] = {128,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0 };
        size_t key_len = 16;

        // data
        uint8_t data[1024] = { 128,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0 };
        size_t data_len = sizeof(data);

        // iv
        uint8_t enc_iv[16] = {128,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0 };

fprintf (stderr, "%16llx\n", *(uint64_t*)data);

        // [F] generate the round keys
        (**enc).init(enc, key, key_len);

        // encryption; happens in-place, i.e. data is overwritten;
        // iv not original afterwards, contains last cipher block
        (**enc).run(enc, enc_iv, data, data_len);

// should deliver the 0EDD33... (endianess!) for (iv XOR data) = 00000... and key = 80000...
fprintf (stderr, "%16llx\n", *(uint64_t*)data);

}
```

Good news: There is only one place _per object_ to choose the cipher implementation [A]. We would probably need three objects: `enc`, `dec` and `enc_iv` [B]. Those [B] and maybe the underlying memory [C1] will probably go along with the cipher-setting [A] to `transform_aes.c`'s global. Variable names might need some polish [A][C1].

As far as I can judge, [C2] and [D] should become part of `n2n_transop_aes_cbc_init(...)`. The object can be used afterwards [E]. 

[D] and [F] look similar and in fact, it is a duplication. Why to repeat? Even though the keys for `enc` and `dec` objects are to be set [F] in `setup_aes_key(...)` anyway, it probably is good practice to have objects initialized [D] as part of their instantiation. Same for `enc_iv` which gets reassigned new keys [F] every round before use. But keep in mind: For example, even `(**enc).block_size` would not carry a valid value prior to initialization.

bearSSL does not perform any padding so presumably not too many changes besides the crypto-calls (as outlined above) required.

Next week, I will try to get through the necessary hash functions. Also, I want to find out about the platform-specific flavors (maybe close to [A]). Assembling those pieces, there should be a sound basis to start actually bringing all this to `transop_aes.c`.

Is that a feasible perspective? Do we still want it? Please let me know what you think.

**Comment by Logan007** [comment]
For demonstration purposes, I got the different elements for __encryption__ a little bit more straight, sorted them and even implemented a kind of auto-platform-selection:

```
#include <stdio.h>
#include <bearssl.h>


// instance of class (object interface, externally providing access to methods and fields)
const br_block_cbcenc_class **enc;

// the actual memory (inner) representation of the object (encryption context internally also provides
// space for round keys, hence the name), 'gen' fits all (e.g. 'big', 'small', ...)
br_aes_gen_cbcenc_keys enc_ctx;


void init (void) {

        // specify class to use (choose cipher implementation, here: deafult to 'big'
        // if no faster platform specific available)
        const br_block_cbcenc_class *vtable_enc;

        if ( (NULL == (vtable_enc =  br_aes_pwr8_cbcenc_get_vtable())) &&
             (NULL == (vtable_enc =  br_aes_x86ni_cbcenc_get_vtable())) )
                       vtable_enc = &br_aes_big_cbcenc_vtable;

        // process of instantiation

        //  connecting interface to actual instance's data structure...
        enc = &enc_ctx.vtable;

        // ... and set default values (properties, pointers to methods) using the class' 'constructor' - otherwise object is useless
        // note: if key and key_len are already known, those could be set here. otherwise: some defaults
        vtable_enc->init(enc, 0, 0);
}


void key_setup (void) {

fprintf (stderr, "block length: %u\n", (**enc).block_size);

        // sizewise, the key array is prepared to take up to 256 bit keys
        uint8_t key[32] = {128,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0 };
        size_t key_len = 16;

        // generate the round keys, this context can be re-used
        (**enc).init(enc, key, key_len);
}


void encrypt (void) {

        // data
        uint8_t data[1024] = { 128,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0 };
        size_t data_len = sizeof(data);

        // iv
        uint8_t enc_iv[16] = {128,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0 };

fprintf (stderr, "%16llx\n", *(uint64_t*)data);

        // encryption; happens in-place, i.e. data is overwritten;
        // iv not original afterwards, contains last cipher block
        (**enc).run(enc, enc_iv, data, data_len);

fprintf (stderr, "%16llx\n", *(uint64_t*)data);
}


int main(int argc, char* argv[]) {

        init();
        // from here on: use the object
        key_setup();
        encrypt();
        // no deinit required as no heap alloc
}
```

Also, I found out about __hashing__ done by bearSSL for which some documentation is [provided](https://bearssl.org/oop.html):

```
size_t hash_data(const br_hash_class *hf, const void *data, size_t len, void *dst) {

        br_hash_compat_context sc;

        hf->init(&sc.vtable);
        sc.vtable->update(&sc.vtable, data, len);
        sc.vtable->out(&sc.vtable, dst);

        return (hf->desc >> BR_HASHDESC_OUT_OFF) & BR_HASHDESC_OUT_MASK;
}

hash_data (&br_sha256_vtable, data, data_len, out);
```

Sorry for posting lengthty code, but I consider this documentary for being more brief in the actual implementation.

The next thing I remember is having provided __pull request__ #242.

The `n2n-benchmark` tool shows the following numbers for __speed__:

CPU | AES Hw | `AES_encrypt_*` | `evp_*` | br `aes_big` | br `aes_x86ni`
:--|:--:|--:|--:|--:|--:
Cortex A53s | ⁿ/ₐ | 49.7 MB/s | 48.4 MB/s | 45.4 MB/s | ⁿ/ₐ
Core i5 M430 | ⁿ/ₐ | 108.4 MB/s | 171.3 MB/s¹ | 99.1 MB/s | ⁿ/ₐ
i7 2860QM | ✓ | 173.8 MB/s | 424.6 MB/s |  134.8 MB/s | 545.3 MB/s
i7 7500U | ✓ | 222.5 MB/s | 745.8 MB/s | 194.4 MB/s | 990.7 MB/s


Even though the `aes_x86ni` implementation of bearSSL clearly beats openSSL's hardware accelerated `evp_*` speedwise, on non-x86 platforms (ARM, x86 without AES-NI²) bearSSL only seems to reach up to roughly 91 percent of openSSLs performance. One of the the reasons could be hardware-specific acceleration for intel without AES-NI (then: MMX, SSE) and ARM-specific optimizations that are still lacking in bearSSL – the author might [add](https://bearssl.org/todo.html#feature-schedule) it later.

So, do we still want it?

¹ openSSL seems to use MMX or SSE in `evp_*` if available in case AES-NI is not present.
² I was not able to test it on POWER8 CPUs.

**Comment by lucaderi** [comment]
I have seen your pull requests and I think we should speak a bit about the future of n2n including
- What SSL library to use
- How to fix some n2n protocol issues to make it really distributed and more robust

Can we schedule a call or public discussion with the community to steer the development of n2n? You can reach me at deri@ntop.org or @lucaderi (telegram). I would like do decide about the features of v3, where we want to go, and who can help with the development.

As of bearSSL, I have looked at the code. The library as @Logan007 said, does not seem to be optimised, the development is stuck since a long time. Instead I see better alternatives https://boringssl.googlesource.com/boringssl/ for instance that is based on OpenSSL (so mostly compatible), supported and a more long-term solution. This will be an argument for discussion.

**Comment by Logan007** [comment]
@lucaderi I have just sent you an eMail (subject: n2n).

**Comment by Logan007** [comment]
Pondering the current options, I guess we will stick with openSSL for now; mainly because it could be considered more reviewed, maintained, and it performs better as seen above.

In special cases, whenever small footprint is a requirement, please see the above examples and pull request #242 (I will close it soon, but it will remain accessible) for how it can be done. I think, for a small project like n2n it is too complicated to maintain two different implementations in parallel.

If anyone knows about some other open source crypto-library – it does not necessarily need to have the SSL part – which performs comparable to or better than openSSL's ciphers (AES, ChaCha) on different platforms, which is portable and has a smaller footprint than openSSL, I will be happy to discuss and maybe even implement. Please just open a new issue.

Unfortunately, the only [comparison](https://panthema.net/2008/0714-cryptography-speedtest-comparison/) I found seems to be a bit outdated...

**Comment by skyformat99** [comment]
That's a great work and description. Thank you

**Comment by Logan007** [comment]
That was a fruitful discussion from which we all have learned a lot, I guess. So, thank you @skyformat99  for bringing it up!

Now, before I _really_ close this file, I just want to leave a final hint on how the freshly proposed ChaCha20 (#230, #235) could be implemented using bearSSL. Just to have all information gathered in one place and just in case somebody would need it for a smaller footprint...

```
        // global (if required)
        // function pointer that will hold the pointer to proper implementation
        br_chacha20_run chacha20;

        // init
        // try to get faster implementation if available, otherwise use default 'ct'
        if ( (NULL == (chacha20 = br_chacha20_sse2_get())) ) {
                chacha20 =  br_chacha20_ct_run;
        }

        // key
        uint8_t key[32] = { 128,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0};

        // iv & counter (12 + 4 bytes)
        uint8_t iv[12]  = { 0,1,2,3,4,5,6,7,8,9,10,11 };
        uint32_t counter = 7;

        // data
        uint8_t data[1024] = { 128,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0 };
        size_t len = 1024;

        // encrypt
        uint32_t new_counter = chacha20 (key, iv, counter, data, len);
```

It seems to work, but I have not performed in-depth testing. As opposed to AES, there is no context but only a function pointer to the en/decryption routine (same). It can always be reused. Also, I am not aware of how this speedwise compares to openSSL.

**Comment by lucaderi** [comment]
Implemented via pull request
